catkin_home/src/action_selectors/scripts/ObjectDetector.py

```diff
@@ -0,0 +1,203 @@
+        # and that will be deleted when closed.
+
+        # Resize the image to meet the requirements of ObjDetecion.
+        # TODO: Check size to see if it will enlarge or shrink and use
+        # CV_INTER_AREA. Check https://docs.opencv.org/2.4/modules/imgproc/doc/geometric_transformations.html#resize
+        # also check whether changing the ratio is very harmful.
+        cv_image = cv2.resize(cv_image, (800, 600), interpolation=cv2.INTER_LINEAR)
+        _, img_buffer = cv2.imencode(".jpg", cv_image)
+
+        temp_img_file.write(img_buffer.tostring())
+        temp_img_file.flush()
+        rospy.loginfo("Tmp image saved (" + temp_img_file.name +"), calling script")
+
+        while True:
+            if obj_det_process.stdout.readline().decode('utf-8') == '~#waiting input#~\n':
+                break
+        obj_det_process.stdin.write((temp_img_file.name + '\n').encode('utf-8'))
+        obj_det_process.stdin.flush()
+        while True:
+            if obj_det_process.stdout.readline().decode('utf-8') == '~#result#~\n':
+                break
+        json_result = obj_det_process.stdout.readline().decode('utf-8')[:-1]
+
+        rospy.loginfo("Results ready")
+
+    #rospy.loginfo("Json result=" + str(json_result))
+    detected_objects = json.loads(json_result)
+    objects = []
+    for object_name in detected_objects:
+        objects.append(
+            ObjectDetected(
+                name=obj,
+                score=detected_objects[object_name]['score'],
+                x_min=detected_objects[object_name]['xmin'],
+                y_min=detected_objects[object_name]['ymin'],
+                x_max=detected_objects[object_name]['xmax'],
+                y_max=detected_objects[object_name]['ymax'],
+            ),
+        )
+    publisher.publish(ObjectsDetected(objects_detected=objects))
+
+def init_obj_det_process():
+    '''
+    Open the object detection script and leave it ready after start up tasks.
+    '''
+    try:
+        process = subprocess.Popen(
+            COMMAND_OBJ_DETECTION,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            # TODO: Check how to send this to null.
+            stderr=subprocess.STDOUT,
+            bufsize=1,
+            env=ENV_VARS_COMMAND,
+            cwd=CWD_COMMAND_OBJ_DETECTION,
+            universal_newlines=True,
+        )
+        if process.returncode != None:
+            return None
+    except:
+        return None
+    rospy.loginfo("Process opened successfully")
+
+    while True:
+        if process.stdout.readline().decode('utf-8') == "~#ready#~\n":
+            break
+
+    rospy.loginfo("Process is ready")
+
+    return process
+
+def close_obj_det_process():
+    rospy.loginfo("Closing process")
+
+    '''
+    This is not the best way because the script callback could be interrupted
+    while it is stuck in another reading.
+    obj_det_process.stdin.write('~#exit#~\n'.encode('utf-8'))
+    obj_det_process.stdin.flush()
```

Comment:
It would be a good idea to call this on node shutdown as a callback!

Reply:
That could help when it is just initializing the script and a cancel comes. Let me try to add it.

Reply:
I found that the callbacks are canceled but the main function continues until the end, but obviously the `spin()` and `sleep()` calls don't block anymore. Then, to avoid closing the subprocess and then having the main reopening it and later closing again, I left it like only printing.

If we really want to kill it completely let's use Ctrl + D instead. Other thing is that inside the shutdown function `exit()` didn't quit the main's execution.